src/app/(potioncraft)/potioncraft/actions.ts

```diff
@@ -0,0 +1,111 @@
+"use server";
+
+import { IngredientSchema, Potion } from "@/types";
+import { prisma } from "@/app/utils/context";
+import { Ingredient, User } from "@prisma/client";
+import { revalidatePath } from "next/cache";
+import {z} from 'zod';
+
+interface SpendIngredientsProps {
+  ingredients: z.infer<typeof IngredientSchema>[]
+}
+
+export const spendIngredients = async ({
+  ingredients,
+}: SpendIngredientsProps) => {
+  for (let ingredient of ingredients) {
+    if (ingredient.quantity === 1) {
+      await prisma.ingredient.delete({
+        where: {
+          id: ingredient.id as string,
+        },
+      });
+      return;
+    }
+    await prisma.ingredient.update({
+      where: {
+        id: ingredient.id as string,
+      },
+      data: {
+        ...ingredient,
+        id: ingredient.id as string,
+        quantity: ingredient.quantity - 1,
+      },
+    });
+  }
+};
+export const increaseIngredient = async ({
+  ingredient,
+  amount,
+}: {
+  ingredient: z.infer< typeof IngredientSchema>;
+  amount: number;
+}) => {
+  const increasedIngredient = await prisma.ingredient.update({
+    where: {
+      id: ingredient.id as string,
+    },
+    data: {
+      quantity: ingredient.quantity + amount,
+    },
+  });
+  revalidatePath(`${process.env.BASE_URL}/potioncraft`);
+  return increasedIngredient;
+};
+
+interface AddPotionToUserProps {
+  userId: User["clerkId"];
+  potion: Potion;
+}
+
+export const addPotionToUser = async ({
+  userId,
+  potion,
+}: AddPotionToUserProps) => {
+  const potionResponse = await prisma.potion.findFirst({
+    where: { userId, name: potion.name },
+  });
+  if (potionResponse === null) {
+    await prisma.potion.create({
+      data: {
+        ...potion,
+        id: undefined,
+        userId,
+      },
+    });
+  } else
+    [
+      await prisma.potion.update({
+        where: {
+          id: potionResponse.id,
+          name: potion.name,
+        },
+        data: {
+          quantity: potionResponse.quantity + 1,
+        },
+      }),
+    ];
+
+  revalidatePath(`${process.env.BASE_URL}/potioncraft`);
+};
+
+interface AddIngredientToUserProps {
+  userId: User["clerkId"];
+  ingredients: Ingredient[];
+}
+
+export const addIngredientsToUser = async ({
+  userId,
+  ingredients,
+}: AddIngredientToUserProps) => {
+  for (let ingredient of ingredients) {
+    await prisma.ingredient.create({
+      data: {
+        ...ingredient,
+        id: undefined,
+        userId,
+      },
+    });
+  }
+  revalidatePath(`${process.env.BASE_URL}/potioncraft`);
+};
```

Comment:
```suggestion
"use server";

import { IngredientSchema, PotionSchema } from "@/types";
import { prisma } from "@/app/utils/context";
import { Ingredient, User } from "@prisma/client";
import { revalidatePath } from "next/cache";
import { z } from 'zod';

const SpendIngredientsSchema = z.object({
  ingredients: z.array(IngredientSchema),
});

const IncreaseIngredientSchema = z.object({
  ingredient: IngredientSchema,
  amount: z.number().int().positive(),
});

const AddPotionToUserSchema = z.object({
  userId: z.string(),
  potion: PotionSchema,
});

const AddIngredientToUserSchema = z.object({
  userId: z.string(),
  ingredients: z.array(IngredientSchema),
});

export const spendIngredients = async (props: z.infer<typeof SpendIngredientsSchema>): Promise<void> => {
  try {
    const { ingredients } = SpendIngredientsSchema.parse(props);

    await prisma.$transaction(async (prisma) => {
      for (const ingredient of ingredients) {
        if (ingredient.quantity === 1) {
          await prisma.ingredient.delete({
            where: { id: ingredient.id },
          });
        } else {
          await prisma.ingredient.update({
            where: { id: ingredient.id },
            data: { quantity: { decrement: 1 } },
          });
        }
      }
    });
  } catch (error) {
    console.error("Error spending ingredients:", error);
    if (error instanceof z.ZodError) {
      throw new Error("Invalid input for spending ingredients");
    }
    throw new Error("Failed to spend ingredients");
  }
};

export const increaseIngredient = async (props: z.infer<typeof IncreaseIngredientSchema>): Promise<Ingredient> => {
  try {
    const { ingredient, amount } = IncreaseIngredientSchema.parse(props);

    const increasedIngredient = await prisma.ingredient.update({
      where: { id: ingredient.id },
      data: { quantity: { increment: amount } },
    });
    revalidatePath(`${process.env.BASE_URL}/potioncraft`);
    return increasedIngredient;
  } catch (error) {
    console.error("Error increasing ingredient:", error);
    if (error instanceof z.ZodError) {
      throw new Error("Invalid input for increasing ingredient");
    }
    throw new Error("Failed to increase ingredient");
  }
};

export const addPotionToUser = async (props: z.infer<typeof AddPotionToUserSchema>): Promise<void> => {
  try {
    const { userId, potion } = AddPotionToUserSchema.parse(props);

    await prisma.$transaction(async (prisma) => {
      const existingPotion = await prisma.potion.findFirst({
        where: { userId, name: potion.name },
      });

      if (existingPotion === null) {
        await prisma.potion.create({
          data: {
            ...potion,
            id: undefined,
            userId,
          },
        });
      } else {
        await prisma.potion.update({
          where: { id: existingPotion.id },
          data: { quantity: { increment: 1 } },
        });
      }
    });

    revalidatePath(`${process.env.BASE_URL}/potioncraft`);
  } catch (error) {
    console.error("Error adding potion to user:", error);
    if (error instanceof z.ZodError) {
      throw new Error("Invalid input for adding potion to user");
    }
    throw new Error("Failed to add potion to user");
  }
};

export const addIngredientsToUser = async (props: z.infer<typeof AddIngredientToUserSchema>): Promise<void> => {
  try {
    const { userId, ingredients } = AddIngredientToUserSchema.parse(props);

    await prisma.ingredient.createMany({
      data: ingredients.map(ingredient => ({
        ...ingredient,
        id: undefined,
        userId,
      })),
    });
    revalidatePath(`${process.env.BASE_URL}/potioncraft`);
  } catch (error) {
    console.error("Error adding ingredients to user:", error);
    if (error instanceof z.ZodError) {
      throw new Error("Invalid input for adding ingredients to user");
    }
    throw new Error("Failed to add ingredients to user");
  }
};
```

# Safety and Performance Improvements

- Added Zod schemas for input validation
- Implemented try-catch blocks with specific error handling
- Used Prisma transactions for multi-step operations
- Optimized database operations (e.g., using `createMany`)
- Improved type safety with TypeScript and Zod